### PR TITLE
Feature/refactor build connectors

### PR DESCRIPTION
**Pull request by Borahm**
⚠️ Fixes: 

## ℹ️ What is the context for these changes?

...

## 🕹️ Demonstration

...

## 🎩 How can this be tophatted?

...

## ✅ Checklist

- [ ] Tested on mobile
- [ ] Tested on multiple browsers
- [ ] Tested for accessibility
- [ ] Includes unit tests
- [ ] Updated relevant documentation for the changes (if necessary)

### CODE REVIEW

File: packages/connect-wallet/src/hooks/useDefaultConnectors.ts [hunk 1]
```diff
@@ -1,24 +0,0 @@
-import {useContext} from 'react';
-
-import {getDefaultConnectors} from '../connectors/getDefaultConnectors';
-import {ConnectWalletContext} from '../providers/ConnectWalletProvider';
-import {Connector, ConnectorInstance} from '../types/connector';
-
-export function useDefaultConnectors() {
```
Comment (jamiely) [inline packages/connect-wallet/src/hooks/useDefaultConnectors.ts:7]: What are the thoughts on whether removing this requires a major update? Although it is not exported from the package index, it is exported from this file.

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx [hunk 2]
```diff
@@ -12,20 +13,39 @@ import {ProviderProps} from './types';
...
   const contextValue: ConnectWalletProviderValue = useMemo(() => {
+    let contextualConnectors = connectors;
+
+    if (!contextualConnectors.length) {
+      const {connectors: defaultConnectors} = buildConnectors({
+        appName: '',
+        chains,
+        includeDefaults: true,
+      });
+      contextualConnectors = defaultConnectors;
```
Comment (jamiely) [inline packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:32]: It seems like this might be similar to what `getDefaultConnectors` does? If so, should it be used here? Also, should `getDefaultConnectors` be refactored to use `buildConnectors`?

File: packages/connect-wallet/src/components/Playground/Playground.stories.tsx [hunk 3]
```diff
@@ -25,18 +24,18 @@ const Template = ({wallets}: {wallets: 'Ethereum' | 'Solana'}) => {
     publicProvider(),
   ]);
 
-  const {connectors} = getDefaultConnectors({chains});
+  const {connectors, wagmiConnectors} = buildConnectors({appName: '', chains});
```
Comment (jamiely) [inline packages/connect-wallet/src/components/Playground/Playground.stories.tsx:27]: It is a little confusing for me seeing `appName` blank. Should we make it optional? It seems only required for coinbase. Or maybe we can just improve the documentation.

File: packages/connect-wallet/src/connectors/buildConnectors.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    MetaMask({chains}),
+    Coinbase({appName, chains}),
+    Rainbow({chains}),
+    LedgerLive({chains}),
+    WalletConnect({chains}),
+  ];
+
+  defaultAvailableConnectors.forEach(
+    ({createConnector, id, ...provided}: ConnectorInstance) => {
+      const createdConnector = createConnector();
```
Comment (caropinzonsilva) [inline packages/connect-wallet/src/connectors/buildConnectors.ts:83]: This can be moved to after the next two `ifs` since it is not used until then. This would improve code readability and  performance since we would only define it when we are going to use it.

File: packages/connect-wallet/src/connectors/buildConnectors.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+         * Context: MetaMask on mobile, Rainbow, etc. will use WalletConnect bridge
+         * and need to use a WC connector. These are all going to use the same WC
+         * connector from Wagmi. But, WalletConnect itself will use two different
+         * WC connectors. One for support with WalletConnect's modal and one without.
+         */
+
+        // When we begin adding more connectors (e.g. Ledger, Rainbow, etc.) we should
+        // take a look at this code again to make sure we're seeing the expected connectors.
+        // We should always have TWO walletConnect connectors (one with a qr code and one without).
+        if (connectors.some((item) => item.id === 'walletConnect')) {
```
Comment (caropinzonsilva) [inline packages/connect-wallet/src/connectors/buildConnectors.ts:109]: I think instead of having nested ifs, it would be better if we define this variable after the `isWalletConnect` with a descriptive name and then do the logic in the same if:
```javascript
const isWalletConnect = createdConnector.id === 'walletConnect';
const descriptiveName =  connectors.some((item) => item.id === 'walletConnect');
if (isWalletConnect && descriptiveName) {
  return;
}
```

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx (hunk 2 above)
Comment (caropinzonsilva) [reply packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:32]: Good question. From [this change](https://github.com/Shopify/blockchain-components/pull/38/files#diff-755ee8ba3e3b0d00dc4e1e8713581e6ad441f004581ca7b525153ff4c7f75952L15) I thought `getDefaultConnectors` would get deleted.

File: packages/connect-wallet/src/hooks/useDefaultConnectors.ts (hunk 1 above)
Comment (QuintonC) [reply packages/connect-wallet/src/hooks/useDefaultConnectors.ts:7]: IMO this shouldn't require a major update.

This was only used internally to access the default connectors which are the same list of connectors returned via `getDefaultConnectors`.

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx (hunk 2 above)
Comment (QuintonC) [reply packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:32]: I think we want to deprecate `getDefaultConnectors`, or instead just have `getDefaultConnectors` call `buildConnectors` internally.

File: packages/connect-wallet/src/components/Playground/Playground.stories.tsx (hunk 3 above)
Comment (QuintonC) [reply packages/connect-wallet/src/components/Playground/Playground.stories.tsx:27]: Good point... I think we might have to overload it since it is required for Coinbase 🤔

**Review by QuintonC — DISMISSED** [review]
I'm converting this PR to a draft and requesting changes. The "request changes" is only to block this from accidentally shipping to main when it's not quite ready.

1. This PR is based off of #32 and that PR is being merged into a feature branch, [`feature/customizable-wallet-list`](https://github.com/Shopify/blockchain-components/tree/feature/customizable-wallet-list).
    - It might be best to wait until that PR is merged into the feature branch before opening the PR for review.
2.  This PR should also be merging to the feature branch mentioned above rather than `main`

**Comment by caropinzonsilva** [comment]
@Borahm @QuintonC I based the PR in the other branch and rebased it so that it only includes the new changes. We might need to do a rebase if @lesliexin force pushes any changes to that other PR, but this will unblock this PR from getting reviewed.

**Comment by lesliexin** [comment]
@caropinzonsilva @Borahm @QuintonC Just force pushed to the other PR addressing some PR comments

File: apps/playground/src/wagmi.ts [hunk 6]
```diff
@@ -12,13 +15,24 @@ const {chains, provider, webSocketProvider} = configureChains(
   ],
 );
 
-const {connectors} = getDefaultConnectors({chains});
+const argentConnector: Connector = {
```
Comment (Borahm) [inline apps/playground/src/wagmi.ts:18]: I will remove this, but I wanted to see if it works!

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx [hunk 7]
```diff
@@ -12,20 +13,38 @@ import {ProviderProps} from './types';
...
   orderAttributionMode = 'required',
   statementGenerator,
   theme,
 }: PropsWithChildren<ProviderProps>) => {
   const contextValue: ConnectWalletProviderValue = useMemo(() => {
+    let contextualConnectors = connectors;
+
+    if (!contextualConnectors.length) {
+      const {connectors: defaultConnectors} = buildConnectors({
+        appName: '',
```
Comment (jamiely) [inline packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:28]: Is appName still required?

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx [hunk 8]
```diff
@@ -12,20 +13,38 @@ import {ProviderProps} from './types';
...
   requireSignature = true,
   orderAttributionMode = 'required',
   statementGenerator,
   theme,
 }: PropsWithChildren<ProviderProps>) => {
   const contextValue: ConnectWalletProviderValue = useMemo(() => {
+    let contextualConnectors = connectors;
+
+    if (!contextualConnectors.length) {
+      const {connectors: defaultConnectors} = buildConnectors({
```
Comment (jamiely) [inline packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:27]: Thoughts on adding a unit test that when `connectors` is empty, `buildConnectors` is called?

**Comment by jamiely** [comment]
I tophatted and, after some code uncommenting, see this:

image

File: packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx (hunk 8 above)
Comment (Borahm) [reply packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:27]: Good call. Will add

Comment (QuintonC) [reply packages/connect-wallet/src/providers/ConnectWalletProvider/provider.tsx:27]: You might need test-infra for this. IMO I'd defer this addition @bo and add this as a part of the larger feature PR prior to merging into main. Perhaps create a task inside the larger PR to add this unit test.

**Comment by QuintonC** [comment]
> I tophatted and, after some code uncommenting, see this:
> 
> image

To clarify, are you asking if this is an unexpected behavior or stating this is expected? Sorry, just trying to understand this comment :) 

This looks like expected behavior to me since Argent is defined as a WalletConnect connector.

File: apps/playground/src/assets/Argent.tsx [hunk 9]
```diff
@@ -0,0 +1,15 @@
+export const Argent = (
```
[the remark this answers is not in the document]
Comment (caropinzonsilva) [reply apps/playground/src/assets/Argent.tsx:1]: I'm just commenting here so that we do not forget to remove it :)

File: packages/connect-wallet/src/components/Playground/Playground.stories.tsx [hunk 10]
```diff
@@ -25,18 +24,18 @@ const Template = ({wallets}: {wallets: 'Ethereum' | 'Solana'}) => {
     publicProvider(),
   ]);
 
-  const {connectors} = getDefaultConnectors({chains});
+  const {connectors, wagmiConnectors} = buildConnectors({chains});
```
Comment (caropinzonsilva) [inline packages/connect-wallet/src/components/Playground/Playground.stories.tsx:27]: This is not a comment for this PR, but just an idea for a future one! We could add a story with custom connectors so that we have somewhere to always test this ✨

**Review by caropinzonsilva — APPROVED** [review]
Code looks good! I 🎩 in Chromatic the previous behaviour and it works as expected: https://639b1f308693132693d9b82c-rodvoyslwc.chromatic.com/?path=/story/connect-wallet--playground&args=wallets:Ethereum

**Review by QuintonC — COMMENTED** [review]
Just a couple of documentation things in the README to address, but otherwise looks solid. Great work 👏

File: packages/connect-wallet/src/components/Playground/Playground.stories.tsx (hunk 10 above)
Comment (Borahm) [reply packages/connect-wallet/src/components/Playground/Playground.stories.tsx:27]: So, I actually have a question about that. Im not super familiar with Storybooks (on my todo list) and was wondering how you made this work: https://639b1f308693132693d9b82c-rodvoyslwc.chromatic.com/?path=/story/connect-wallet--playground&args=wallets:Ethereum I could not make it work on my local machine and I always got an error.

Aside of that, Im happy to add one, just not 100% sure how to go about it

File: packages/connect-wallet/src/types/connector.ts [hunk 11]
```diff
@@ -65,6 +65,10 @@ export type Connector = Omit<ConnectorInstance, 'createConnector'> & {
   ready?: boolean;
 };
 
+export type CustomConnector = Omit<ConnectorInstance, 'createConnector'> & {
```
[the remark this answers is not in the document]
Comment (Borahm) [reply packages/connect-wallet/src/types/connector.ts:68]: nice!

File: packages/connect-wallet/src/components/Playground/Playground.stories.tsx (hunk 10 above)
Comment (caropinzonsilva) [reply packages/connect-wallet/src/components/Playground/Playground.stories.tsx:27]: There was an error in that story, but it should be fixed after the Storybook v7 change :)
